chapters/operatorsandexpressions.tex

```diff
@@ -1479,6 +1469,21 @@ \subsection{Discrete-Time Expressions}\label{discrete-time-expressions}
 functions do not become active between events.
 \end{nonnormative}
 
+For a scalar equation \lstinline!expr1 = expr2! where neither expression is of base type \lstinline!Real!, both expressions must be discrete-time expressions.
+For a record equation, the same rule applies to the scalar elements of the equation.
```

Comment:
And similarly
```suggestion
For a record equation, the same rule applies to the non-record elements of the equation.
```

Reply:
I prefer the original version that said the record is first split into basic types. Either "scalar" or "non-record elements" to me leave questions about what happens to arrays or subrecords. Was there something wrong with the original formulation?

Reply:
One problem I saw with "basic types" is that _basic type_ doesn't seem to be an established concept.

Note that giving a rule for the scalar elements of the equation solves the problem of what happens to sub-records.  Alternatively, a formulation with _the rule applies recursively_ should take care of any sub-records.

However, the problem to solve first is probably to just make sure that we agree on what the rule should say for array equations.  Consider these equations:
```
  Boolean x;
  Boolean y;
equation
  /* Equation a: */
  {x, y} = {time > 0.5, time > 0.6};
  /* Equation b: */
  {x, y} = {noEvent(time > 0.5), time > 0.6};
  /* Equation c: */
  {x, y} = noEvent({time > 0.5, time > 0.6});
```

I suppose the intention is that the rule operates on the entire LHS and RHS arrays?  That is, the variability of the LHS is discrete-time in all equations, but only (a) has a discrete-time RHS and satisfies the rule.

For my own sake, I'll also include the record variant for clarity:
```
record R
  Boolean x;
  Real y;
end R;

  R r;
equation
  /* Equation a: */
  r = R(time > 0.5, time);
  /* Equation b: */
  r = R(noEvent(time > 0.5), time);
  /* Equation c: */
  r = noEvent(R(time > 0.5, time));
```

Here, the fact that the entire RHS is continuous-time in all cases doesn't violate the rule; each record member needs to be considered separately.  But what if the RHS is given by an intricate function?
```
function foo
  input Boolean p;
  input Real t;
  output R r;
algorithm
  /* Complicated mess. */
end foo;

  /* Equation d: */
  r = foo(time > 0.5, time);
```

One advantage of presenting the rule non-normatively and instead show how it can be inferred is that we wouldn't have to go into details like this here; it would be sufficient to let the other parts of the specification define when `r.x` is considered solvable in the equation.

Reply:
> One problem I saw with "basic types" is that _basic type_ doesn't seem to be an established concept.
> 
> Note that giving a rule for the scalar elements of the equation solves the problem of what happens to sub-records. Alternatively, a formulation with _the rule applies recursively_ should take care of any sub-records.

Yes, that seems fine.

> However, the problem to solve first is probably to just make sure that we agree on what the rule should say for array equations. Consider these equations:
> 
> ```
>   Boolean x;
>   Boolean y;
> equation
>   /* Equation a: */
>   {x, y} = {time > 0.5, time > 0.6};
>   /* Equation b: */
>   {x, y} = {noEvent(time > 0.5), time > 0.6};
>   /* Equation c: */
>   {x, y} = noEvent({time > 0.5, time > 0.6});
> ```
> 
> I suppose the intention is that the rule operates on the entire LHS and RHS arrays? That is, the variability of the LHS is discrete-time in all equations, but only (a) has a discrete-time RHS and satisfies the rule.

Yes, arrays seen as one entity, because arrays need to be homogenous. 
Similarly as the entire RHS is Real below (and thus this example is illegal):
```
  Real x;
  Integer y;
equation
{x,y}={time,2};
```
Whereas records members are different.

> For my own sake, I'll also include the record variant for clarity:
> 
> ```
> record R
>   Boolean x;
>   Real y;
> end R;
> 
>   R r;
> equation
>   /* Equation a: */
>   r = R(time > 0.5, time);
>   /* Equation b: */
>   r = R(noEvent(time > 0.5), time);
>   /* Equation c: */
>   r = noEvent(R(time > 0.5, time));
> ```
> 
> Here, the fact that the entire RHS is continuous-time in all cases doesn't violate the rule; each record member needs to be considered separately. But what if the RHS is given by an intricate function?

That is more complicated, but any function call will only return a discrete-time valued expression if all inputs are discrete-time valued expression; and thus a record with non-Reals returned by a function called with non discrete-time expression will not work as rhs.

Reply:
> Yes, arrays seen as one entity, because arrays need to be homogenous.
> Similarly as the entire RHS is Real below (and thus this example is illegal):

OK, that's a good example pinpointing something I've missed.

> That is more complicated, but any function call will only return a discrete-time valued expression if all inputs are discrete-time valued expression; and thus a record with non-Reals returned by a function called with non discrete-time expression will not work as rhs.

Does this mean that a record constructor is not considered a function in this sense – that a record returned by a non-constructor function has a variability (determined from the function arguments), whereas a record constructor expression doesn't have a variability?

Reply:
It's been change to _non-record_ now, and I've also tried to elaborate what's going on in the non-normative text.  Maybe we can restart the discussion based on the new non-normative text?

I also added a new example, and moved the existing example out of the discrete-time-specific section.

Reply:
It seems this conversation is becoming obsolete in view of separate comments on this issue further down.  Maybe it's time to resolve, and if needed restart based on the current formulations?

Reply:
I think this is now outdated and can be resolved. Resolving.